torch/nn/utils/_stateless.py

```diff
@@ -0,0 +1,56 @@
+import contextlib
+
+import torch
+
+
+def _get_module_parameters_and_buffers(module):
+    """Helper function that returns a dictionary
+    with only parameters and buffers, we don't use state_dict as it might
+    be polluted with module attributes.
+    """
+    parameters_and_buffers = dict(module.named_parameters())
+    parameters_and_buffers.update(dict(module.named_buffers))
+    # TODO: clean entries that are already parameterized?
+    # params with weight norm/spectral norm applied
+    return parameters_and_buffers
+
+@contextlib.contextmanager
+def reparametrize_module(module, parameters_and_buffers):
+    # Parametrization does not support to change submodules directly
+    for name, tensor in parameters_and_buffers.items():
+        _reparametrize_in_submodule(module, name.split("."), tensor)
+    yield
+    for name in parameters_and_buffers:
+        _remove_reparametrize_in_submodule(module, name.split("."))
+
+class _ReparametrizedTensor(torch.nn.Module):
+    def __init__(self, tensor):
+        super().__init__()
+        self._tensor = tensor
+
+    def forward(self, original):
+        return self._tensor
+
+def _reparametrize_in_submodule(module, path, tensor):
+    if len(path) == 1:
+        # We should be careful as the current API does not allow to reparametrize
```

Comment:
The problem I saw here, is when we delete the reparametrization the first time, it wiped all the metadata so the second time had issues, but maybe I spoke too fast, I will recheck it.

Reply:
```python
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self._l = torch.nn.Linear(1, 1)
        self.register_buffer('buff', torch.tensor([2]))

    def forward(self, x):
        return self._l(x)

m = MyModule()
with torch.no_grad():
    m._l.weight[:] = 1
    m._l.bias[:] = 0
np = torch.tensor([4])
print(m.state_dict())
print(dict(m.named_buffers()))
torch.nn.utils.parametrize.register_parametrization(m, 'buff', Replacer(np))
print(dict(m.named_buffers()))
np = torch.tensor([5])
torch.nn.utils.parametrize.register_parametrization(m, 'buff', Replacer(np))
print(dict(m.named_buffers()))
torch.nn.utils.parametrize.remove_parametrizations(m, 'buff', False)
print(dict(m.named_buffers()))
torch.nn.utils.parametrize.remove_parametrizations(m, 'buff', False)
print(dict(m.named_buffers()))
```

output 

```
OrderedDict([('buff', tensor([2])), ('_l.weight', tensor([[1.]])), ('_l.bias', tensor([0.]))])
{'buff': tensor([2])}
{'parametrizations.buff.original': tensor([2])}
{'parametrizations.buff.original': tensor([2])}
{'buff': tensor([2])}
Traceback (most recent call last):
  File "tpar.py", line 52, in <module>
    torch.nn.utils.parametrize.remove_parametrizations(m, 'buff', False)
  File "/home/ecastill/em-pytorch/torch/nn/utils/parametrize.py", line 601, in remove_parametrizations
    raise ValueError(f"Module {module} does not have a parametrization on {tensor_name}")
ValueError: Module MyModule(
  (_l): Linear(in_features=1, out_features=1, bias=True)
) does not have a parametrization on buff
```

Reply:
I confirmed it worked fine, I was doing some weird stuff :),

However, it seems calling `remove_reparametrizations` cleans all at the same time, so this may cause problems if we pass an already parametrized models, because the original ones will be wiped.

Reply:
But the thing is that here, the parametrization that you add is not on 'buff' but 'parametrizations.buff.original' and so when you remove the parametrization on 'parametrizations.buff.original', you don't modify the user one on 'buf'.
Or I am missing something?

Reply:
Let me show you what happens when we pass a module parametrized with spectral normalization.

```python

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(1, 1)

    def forward(self, x):
        return self.l1(x)

mod = MyModule()
# 1.
print(dict(mod.named_parameters()))
torch.nn.utils.parametrizations.spectral_norm(mod.l1)
# 2. Here we will see the parameterization in effect
print(dict(mod.named_parameters()))
x = torch.rand((1, 1))
parameters = {"l1.weight": torch.nn.Parameter(torch.tensor([[1.0]])),
              "l1.bias": torch.nn.Parameter(torch.tensor([0.0]))}
res = torch.nn.utils._stateless.functional_call(mod, parameters, x)
# 3. parametrization is now gone!
print(dict(mod.named_parameters()))
```

output

```
{'l1.weight': Parameter containing:
tensor([[0.2653]], requires_grad=True), 'l1.bias': Parameter containing:
tensor([0.3915], requires_grad=True)}

{'l1.bias': Parameter containing:
tensor([0.3915], requires_grad=True), 'l1.parametrizations.weight.original': Parameter containing:
tensor([[0.2653]], requires_grad=True)}

{'l1.weight': Parameter containing:
tensor([[0.2653]], requires_grad=True), 'l1.bias': Parameter containing:
tensor([0.3915], requires_grad=True)}
```

Reply:
Ho something is wrong here. We should fix that. (and test it)
Here we should be nesting the parametrization and only removing the "bottom one".

Reply:
Yay! that's what I was trying to say haha, sorry for my bad explanations 😣

Reply:
I will try to fix this

Reply:
The problem above is that you provide a key that maps to something that is not an actual parameter/buffer. That should be forbidden.
Making the key match what `named_parameters()` gives you "l1.parametrizations.weight.original" should work though right?